src/include/btree_cmp_inline.h

```diff
@@ -21,6 +21,152 @@
 #define WT_ALIGNED_16(p) (((uintptr_t)(p)&0x0f) == 0)
 #define WT_VECTOR_SIZE 16 /* chunk size */
 
+#ifndef HAVE_X86INTRIN_H
+/*
+ * __lex_compare_ge_16 --
+ *     Lexicographic comparison routine for data greater than or equal to 16 bytes. Returns: < 0 if
+ *     user_item is lexicographically < tree_item, = 0 if user_item is lexicographically =
+ *     tree_item, > 0 if user_item is lexicographically > tree_item. We use the names "user" and
+ *     "tree" so it's clear in the btree code which the application is looking at when we call its
+ *     comparison function.
+ */
+static WT_INLINE int
+__lex_compare_ge_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    struct {
+        uint64_t a, b;
+    } udata, tdata;
+    uint64_t u64, t64;
+    const uint8_t *uendp, *userp, *tendp, *treep;
+    bool firsteq;
+
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /* skip 16 matching bytes at a time, starting at first possible difference. */
+    for (userp = ustartp, treep = tstartp; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        memcpy(&udata, userp, WT_VECTOR_SIZE);
+        memcpy(&tdata, treep, WT_VECTOR_SIZE);
+        if (udata.a != tdata.a || udata.b != tdata.b)
+            goto final128;
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bound.
+     */
+    memcpy(&udata, uendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+    memcpy(&tdata, tendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+
+final128:
+    firsteq = udata.a == tdata.a;
+    u64 = firsteq ? udata.b : udata.a;
+    t64 = firsteq ? tdata.b : tdata.a;
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    u64 = __builtin_bswap64(u64);
+    t64 = __builtin_bswap64(t64);
+#endif
+
+    return (u64 < t64 ? -1 : u64 > t64 ? 1 : lencmp);
+}
+
+/*
+ * __lex_compare_lt_16 --
+ *     Lexicographic comparison routine for data less than 16 bytes. Returns: < 0 if user_item is
+ *     lexicographically < tree_item, = 0 if user_item is lexicographically = tree_item, > 0 if
+ *     user_item is lexicographically > tree_item. We use the names "user" and "tree" so it's clear
+ *     in the btree code which the application is looking at when we call its comparison function.
+ */
+static WT_INLINE int
+__lex_compare_lt_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    uint64_t ta, tb, ua, ub, u64, t64;
+    const uint8_t *tendp, *uendp;
+
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+    if (len & sizeof(uint64_t)) {
```

Comment:
Sorry @quchenhao, to clarify I meant something like
```
if (len & sizeof(uint64_t)) {        /* len >= 64 bits. len is implicitly less than 128bits since the function accepts 16 bytes or less. */
} else if (len & sizeof(uint32_t)) { /* len >= 32 bits */
} else if (len & sizeof(uint16_t)) { /* len >= 16 bits */
```